node-graph/gcore/src/graphic_element/renderer.rs

```diff
@@ -0,0 +1,188 @@
+	pub culling_bounds: Option<[DVec2; 2]>,
+	pub thumbnail: bool,
+}
+
+impl RenderParams {
+	pub fn new(view_mode: crate::vector::style::ViewMode, culling_bounds: Option<[DVec2; 2]>, thumbnail: bool) -> Self {
+		Self { view_mode, culling_bounds, thumbnail }
+	}
+}
+
+fn format_transform_matrix(transform: DAffine2) -> String {
+	transform.to_cols_array().iter().map(ToString::to_string).collect::<Vec<_>>().join(",")
+}
+
+pub trait GraphicElementRendered {
+	fn render_svg(&self, render: &mut SvgRender, render_params: &RenderParams);
+	fn bounding_box(&self, transform: DAffine2) -> Option<[DVec2; 2]>;
+}
+
+impl GraphicElementRendered for GraphicGroup {
+	fn render_svg(&self, render: &mut SvgRender, render_params: &RenderParams) {
+		self.iter().for_each(|element| element.graphic_element_data.render_svg(render, render_params))
+	}
+	fn bounding_box(&self, transform: DAffine2) -> Option<[DVec2; 2]> {
+		self.iter().filter_map(|element| element.graphic_element_data.bounding_box(transform)).reduce(combine_bounds)
+	}
+}
+
+impl GraphicElementRendered for VectorData {
+	fn render_svg(&self, render: &mut SvgRender, render_params: &RenderParams) {
+		let layer_bounds = self.bounding_box().unwrap_or_default();
+		let transfomed_bounds = self.bounding_box_with_transform(render.transform).unwrap_or_default();
+
+		render.svg.push("<path d=\"".into());
+		let mut path = String::new();
+		for subpath in &self.subpaths {
+			let _ = subpath.subpath_to_svg(&mut path, self.transform * render.transform);
+		}
+		render.svg.push(path.into());
+		render.svg.push("\"".into());
+
+		let style = self.style.render(render_params.view_mode, &mut render.svg_defs, render.transform, layer_bounds, transfomed_bounds);
+		render.svg.push(style.into());
+		render.svg.push("/>".into());
+	}
+	fn bounding_box(&self, transform: DAffine2) -> Option<[DVec2; 2]> {
+		self.bounding_box_with_transform(self.transform * transform)
+	}
+}
+
+impl GraphicElementRendered for Artboard {
+	fn render_svg(&self, render: &mut SvgRender, render_params: &RenderParams) {
+		self.graphic_group.render_svg(render, render_params)
+	}
+	fn bounding_box(&self, transform: DAffine2) -> Option<[DVec2; 2]> {
+		let artboard_bounds = self.bounds.map(|[a, b]| (transform * Quad::from_box([a.as_dvec2(), b.as_dvec2()])).bounding_box());
+		[self.graphic_group.bounding_box(transform), artboard_bounds].into_iter().flatten().reduce(combine_bounds)
+	}
+}
+
+impl GraphicElementRendered for ImageFrame<Color> {
+	fn render_svg(&self, render: &mut SvgRender, _render_params: &RenderParams) {
+		let transform: String = format_transform_matrix(self.transform * render.transform);
+		render
+			.svg
+			.push(format!(r#"<image width="1" height="1" preserveAspectRatio="none" transform="matrix({transform})" href=""#).into());
+		let uuid = generate_uuid();
+		render.svg.push(SvgSegment::BlobUrl(uuid));
+		render.svg.push("\" />".into());
+		render.image_data.push((uuid, self.image.clone()))
```

Comment:
ChatGPT recommends this, is it applicable?

![image](https://github.com/GraphiteEditor/Graphite/assets/4388688/5bcbd567-0785-4baf-8d70-0fc08b944634)